opensquirrel/default_gates.py

```diff
@@ -1,117 +1,211 @@
+        super().__init__(parameter)
+
+    def __call__(self, q: QubitLike) -> BlochSphereRotation:
+        return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=math.pi / 2, phase=0)
 
 
 @named_gate
-def Sdag(q: QubitLike) -> BlochSphereRotation:
-    return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=-math.pi / 2, phase=0)
+class Sdag:
+    def __init__(self, parameter: Any = None):
+        pass
+
+    def __call__(self, q: QubitLike) -> BlochSphereRotation:
+        return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=-math.pi / 2, phase=0)
 
 
 @named_gate
-def T(q: QubitLike) -> BlochSphereRotation:
-    return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=math.pi / 4, phase=0)
+class T(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, q: QubitLike) -> BlochSphereRotation:
+        return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=math.pi / 4, phase=0)
 
 
 @named_gate
-def Tdag(q: QubitLike) -> BlochSphereRotation:
-    return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=-math.pi / 4, phase=0)
+class Tdag(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, q: QubitLike) -> BlochSphereRotation:
+        return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=-math.pi / 4, phase=0)
 
 
 @named_gate
-def Rx(q: QubitLike, theta: Float) -> BlochSphereRotation:
-    return BlochSphereRotation(qubit=q, axis=(1, 0, 0), angle=theta.value, phase=0)
+class Rx(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, q: QubitLike) -> BlochSphereRotation:
+        self.theta = Float(self.parameter).value
+        return BlochSphereRotation(qubit=q, axis=(1, 0, 0), angle=self.theta, phase=0)
 
 
 @named_gate
-def Ry(q: QubitLike, theta: Float) -> BlochSphereRotation:
-    return BlochSphereRotation(qubit=q, axis=(0, 1, 0), angle=theta.value, phase=0)
+class Ry(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, q: QubitLike) -> BlochSphereRotation:
+        self.theta = Float(self.parameter).value
+        return BlochSphereRotation(qubit=q, axis=(0, 1, 0), angle=self.theta, phase=0)
 
 
 @named_gate
-def Rz(q: QubitLike, theta: Float) -> BlochSphereRotation:
-    return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=theta.value, phase=0)
+class Rz(NamedGateFunctor):
+    def __init__(self, parameter: Any = None):
+        super().__init__(parameter)
+
+    def __call__(self, q: QubitLike) -> BlochSphereRotation:
+        self.theta = Float(self.parameter).value
+        return BlochSphereRotation(qubit=q, axis=(0, 0, 1), angle=self.theta, phase=0)
 
 
 @named_gate
-def CNOT(control: QubitLike, target: QubitLike) -> ControlledGate:
-    return ControlledGate(control, X(target))
+class CNOT(NamedGateFunctor):
```

Comment:
If gates are now classes, would the matrix representation of the gate be contained in the class itself?

Reply:
It could.